### PR TITLE
fix: Handling of column types for Presto, Trino, et al.

### LINKED ISSUE #25636: [sqllab] Tableschemaview fails to load with presto using parquet format files
### Summary
On Superset version 3.1.0 the table schema previews fail to load on `sqllab/` and `dataset/add/` paths when using presto datasources and parquet format files.

These were working for Superset version 1.5.3, but have not worked since version 2.0.1 and up to 3.1.0.

<ins>**Error condition** on Superset version 3.1.0 with presto table using `format=PARQUET`</ins>

### Conditions
* This only occurs when those tables are using Presto and a table `format = 'PARQUET'`
* Other database types such as Druid and MySQL do not exhhibit this behaviour
* This occurs when logged in as an Admin user, therefore I believe that it is likely unrelated to #25451
* When the presto table is using `format = 'TEXTFILE'` this error does not occur

### How to reproduce the bug

1. Go to Superset version 3.1.0 as an Admin user and navigate to either of:
  a) `/sqllab`
  b) `/dataset/add`
3. Select a database that uses the `presto` connector type
4. Select any schema
5. Select any table that uses a `format = 'PARQUET'`

### Expected results

I would expect the left hand column to be populated with the column names from the selected schema.

### Actual results

Several error messages appear stating that there were errors fetching table metadata and the left-hand column is not populated.

#### Error messages in the server log

There are **no relevant error messages** in the server log.

We can see the pyhive presto command going through:
```
INFO:pyhive.presto:SHOW COLUMNS FROM "wmf"."aqs_hourly"
INFO:pyhive.presto:SHOW COLUMNS FROM "wmf"."aqs_hourly"
INFO:pyhive.presto:SHOW COLUMNS FROM "wmf"."aqs_hourly"
INFO:pyhive.presto:SELECT * FROM wmf."aqs_hourly$partitions"
ORDER BY year DESC, month DESC, day DESC, hour DESC
LIMIT 1
```
We have a lot of `DEBUG` level messages from `requests_kerberos.kerberos_` and `urllib3.connectionpool` and `spnego._gss` while the request is authenticated and processed, but these appear to show a 401 followed by a successful 200 response.

There are no stack traces shown.

#### Additional Screenshots

<ins>**Error condition** on Superset version 2.1.1 with presto table using `format=PARQUET`</ins>

<ins>**No error** on Superset version 1.5.3 for the same table</ins>

<ins>**No error** on Superset version 2.1.1 with presto table using `format=TEXTFILE`</ins>

### Environment

- browser type and version: Firefox 118.0.1 (64-bit) on Linux, but this affects other browsers.
- superset version: `3.1.0`
- python version: `3.9.2`
- node.js version: `16`
- any feature flags active:
  - `ENABLE_TEMPLATE_PROCESSING`
  - `DASHBOARD_NATIVE_FILTERS`
  - `ENABLE_FILTER_BOX_MIGRATION`
- metadata caching, memcached
 
Metadata database: MariaDB 10.4
Presto version 0.283

Optional components:
```
pyhive[kerberos,presto]==0.7.0
gunicorn[gevent]
apache-superset[hive,presto,mysql,druid,trino,spark,postgres]
pylibmc==1.6.1
```
### Checklist

Make sure to follow these steps before submitting your issue - thank you!

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

### LINKED ISSUE #25962: Failure on creation of a new Presto dataset due to failure to load columns
We have a Presto database connection working fine with virtual datasets (added throw SQLLab).
When trying to add a new dataset of a physical table we're getting an error.

### Reproducing the bug:

1. Go to Datasets
2. Click on '+ Dataset'
3. Select the a Presto DB
4. Select the Schema
5. Select the Table
7. See error

### Expected results

See the table columns on the right pane

### Actual results

An error message appears on the right pane:
"An Error Occurred
Unable to load columns for the selected table. Please select a different table."

Couldn't find any errors in the Superset app container log or any other container.

#### Screenshots

image

### Environment

- Google Chrome, Version 119.0.6045.123
- Superset version: 3.0.1
- python version: 3.11.4
- node.js version: 20.9.0
- any feature flags active: none

### Checklist

- [+ ] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [ +] I have reproduced the issue with at least the latest released version of superset.
- [ +] I have checked the issue tracker for the same issue and I haven't found one similar.

### LINKED ISSUE #26740: [sqllqb] Superset always fires "show create view" although it is a table which results in "error while fetching table metadata"
### Bug description

Superset version : 3.0.3

Error in show table schema / fetching metadata
Superset always fires "show create view" although it is a table which results in "error while fetching table metadata"

### How to reproduce the bug

1. Go to sqllab.
2. select database
3. Select schema
4. Select a table (not view)

### Screenshots/recordings

Screenshot:
image

UI Screenshot:
image

### Superset version

master / latest-dev

### Python version

3.11

### Node version

Not applicable

### Browser

Chrome

### Additional context

_No response_

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

### PR DESCRIPTION
**Pull request by john-bodley**
### SUMMARY

This is a rehash of @brouberol's https://github.com/apache/superset/pull/26782 which was later reverted in https://github.com/apache/superset/pull/28613 as it was problematic with Trino, given both the `PrestoEngineSpec` and `TrinoEngineSpec` derive from the `PrestoBaseEngineSpec`.  

The TL;DR is typically most SQLAlchemy dialects (including Trino) use native SQLAlchemy types for encoding column types, whereas PyHive is unconventional and uses `str`. I asked a similar question [years ago](https://github.com/dropbox/PyHive/issues/121) and you can see [here](https://github.com/dropbox/PyHive/blob/d4ae481675ac5588ba9101596fa26f22ef0e77c4/pyhive/tests/test_presto.py#L52-L81) how these types are materialized.

I hypothesize that part of the confusion when @brouberol's was authoring their PR was that the `ResultSetColumnType` `TypedDict` was incorrect, alluding to the fact that the `type` field was `Optional[str]` as opposed to `Optional[Union[str, TypeEngine, type[TypeEngine]]`. This PR simply ensures that in the `PrestoBaseEngineSpec` class—which handles Hive, Presto, and Trino—that we only try to coerce `str` types to a SQLAlchemy type (if available).

### TESTING INSTRUCTIONS

Added unit tests.

### ADDITIONAL INFORMATION

- [x] Has associated issue: 
  - Fixes https://github.com/apache/superset/issues/25636
  - Fixes https://github.com/apache/superset/issues/25962
  - Fixes https://github.com/apache/superset/issues/26740
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: tests/unit_tests/db_engine_specs/test_presto.py [hunk 1]
```diff
@@ -116,46 +115,44 @@ def test_get_schema_from_engine_params() -> None:
...
-    result = spec.where_latest_partition(
-        mock.MagicMock(),
-        Table("table"),
-        query,
-        columns,
-    )
-    assert result is not None
-    actual = result.compile(
-        dialect=PrestoDialect(), compile_kwargs={"literal_binds": True}
+    assert (
```
Comment (john-bodley) [inline tests/unit_tests/db_engine_specs/test_presto.py:134]: Same logic as before just condensed.

**Comment by admsev** [comment]
@gooroodev review please

**Comment by admsev** [comment]
@gooroodev review

**Comment by gooroodev** [comment]
### Summary of Changes
1. **Presto Engine Specification**:
   - Minor formatting adjustments.
   - Updated the `get_children` function to cast the column type to a string.
   - Modified the `where_latest_partition` function to handle column types that are strings by converting them to SQLAlchemy types if possible.

2. **Superset Typing**:
   - Introduced a new type alias `SQLType` for Union of `TypeEngine` and `type[TypeEngine]`.
   - Updated `ResultSetColumnType` to use the new `SQLType` alias.

3. **Unit Tests for Presto**:
   - Adjusted the `test_where_latest_partition` to use string representations of column types instead of SQLAlchemy types.

4. **Unit Tests for Trino**:
   - Added a new test `test_where_latest_partition` for Trino engine specification.
   - Used `MagicMock` instead of `Mock` for database mocking.
   - Included the Trino dialect for SQL compilation in the test.

### Issues, Bugs, and Typos
1. **File: `superset/db_engine_specs/presto.py`**
   - **Issue**: Redundant import of `annotations` from `__future__`.
   - **Improvement**: Remove the import statement if it's not being used.

2. **File: `superset/superset_typing.py`**
   - **Issue**: The `SQLType` alias is defined but not used in `ResultSetColumnType`.
   - **Improvement**: Ensure `SQLType` is utilized correctly in the `ResultSetColumnType` definition.

3. **File: `tests/unit_tests/db_engine_specs/test_presto.py`**
   - **Issue**: The test function `test_where_latest_partition` has inconsistent parameter typing.
   - **Improvement**: Update the parameter typing to match the expected types.

4. **File: `tests/unit_tests/db_engine_specs/test_trino.py`**
   - **Issue**: The `test_where_latest_partition` function name conflicts with the one in `test_presto.py`.
   - **Improvement**: Rename the test function to avoid conflicts and improve clarity.

### Proposed Code Improvements
1. **Remove Redundant Import in Presto Engine Spec**
   ```python
   # Remove the following line if not used:
   from __future__ import annotations
   ```

2. **Ensure Correct Typing in Superset Typing**
   ```python
   # Ensure SQLType is used correctly
   class ResultSetColumnType(TypedDict):
       name: str  # legacy naming convention keeping this for backwards compatibility
       column_name: str
       type: Optional[Union[SQLType, str]]
       is_dttm: Optional[bool]
       type_generic: NotRequired[Optional["GenericDataType"]]
   ```

3. **Update Typing in Presto Test**
   ```python
   @pytest.mark.parametrize(
       ["column_type", "column_value", "expected_value"],
       [
           ("DATE", "2023-05-01", "DATE '2023-05-01'"),
           ("TIMESTAMP", "2023-05-01", "TIMESTAMP '2023-05-01'"),
           ("VARCHAR", "2023-05-01", "'2023-05-01'"),
           ("INT", 1234, "1234"),
       ],
   )
   def test_where_latest_partition(
       mock_latest_partition: Mock,
       column_type: str,
       column_value: Any,
       expected_value: str,
   ) -> None:
       from superset.db_engine_specs.presto import PrestoEngineSpec
       
       mock_latest_partition.return_value = (["partition_key"], [column_value])
       
       assert (
           str(
               PrestoEngineSpec.where_latest_partition(  # type: ignore
                   database=mock.MagicMock(),
                   table=Table("table"),
                   query=sql.select(text("* FROM table")),
                   columns=[
                       {
                           "column_name": "partition_key",
                           "name": "partition_key",
                           "type": column_type,
                           "is_dttm": False,
                       }
                   ],
               ).compile(
                   dialect=PrestoDialect(),
                   compile_kwargs={"literal_binds": True},
               )
           )
           == f"""SELECT * FROM table \nWHERE "partition_key" = {expected_value}"""
       )
   ```

4. **Rename Trino Test Function**
   ```python
   @patch("superset.db_engine_specs.trino.TrinoEngineSpec.latest_partition")
   @pytest.mark.parametrize(
       ["column_type", "column_value", "expected_value"],
       [
           (types.DATE(), "2023-05-01", "DATE '2023-05-01'"),
           (types.TIMESTAMP(), "2023-05-01", "TIMESTAMP '2023-05-01'"),
           (types.VARCHAR(), "2023-05-01", "'2023-05-01'"),
           (types.INT(), 1234, "1234"),
       ],
   )
   def test_trino_where_latest_partition(
       mock_latest_partition: Mock,
       column_type: SQLType,
       column_value: Any,
       expected_value: str,
   ) -> None:
       from superset.db_engine_specs.trino import TrinoEngineSpec
       
       mock_latest_partition.return_value = (["partition_key"], [column_value])
       
       assert (
           str(
               TrinoEngineSpec.where_latest_partition(  # type: ignore
                   database=MagicMock(),
                   table=Table("table"),
                   query=sql.select(text("* FROM table")),
                   columns=[
                       {
                           "column_name": "partition_key",
                           "name": "partition_key",
                           "type": column_type,
                           "is_dttm": False,
                       }
                   ],
               ).compile(
                   dialect=TrinoDialect(),
                   compile_kwargs={"literal_binds": True},
               )
           )
           == f"""SELECT * FROM table \nWHERE partition_key = {expected_value}"""
       )
   ```

### General Review of Code Quality and Style
- **Code Quality**: The code is well-structured and follows good practices for handling different column types. The use of type hints and `TypedDict` enhances readability and type safety.
- **Style**: The code adheres to PEP 8 standards. The use of type casting and conditional assignments is clear and concise.
- **Testing**: The tests are comprehensive and cover various scenarios for handling column types. However, function naming conflicts should be avoided to ensure clarity.
- **Documentation**: The code could benefit from additional inline comments explaining the reasoning behind certain type conversions and handling mechanisms.

Overall, the pull request effectively addresses the handling of column types for Presto and Trino, with minor improvements needed for clarity and consistency.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)

**Comment by gooroodev** [comment]
### 1. Summary of Changes

The pull request addresses the handling of column types for Presto and Trino engines in the Superset project. The key changes include:

- **Presto Engine:**
  - Handling of column types in the `get_children` and `where_latest_partition` methods.
  - Added logic to convert string column types to SQLAlchemy types dynamically.
- **Trino Engine:**
  - Added a similar handling of column types in the `where_latest_partition` method.
  - Added unit tests for the `where_latest_partition` method.
- **Typing Changes:**
  - Introduced a new type alias `SQLType` in `superset_typing.py` to represent SQLAlchemy column types.
  - Updated `ResultSetColumnType` to use `SQLType`.

### 2. Issues, Bugs, or Typos

#### Issue 1:
- **File:** `superset/db_engine_specs/presto.py`
- **Code:**
  ```python
  match = pattern.match(cast(str, column["type"]))
  ```
  **Problem:** The use of `cast(str, column["type"])` is unnecessary if `column["type"]` is already a string.
  **Improvement:**
  ```python
  match = pattern.match(column["type"])
  ```

#### Issue 2:
- **File:** `superset/db_engine_specs/presto.py`
- **Code:**
  ```python
  if isinstance(col_type, str):
      if col_type_class := getattr(types, col_type, None):
          col_type = col_type_class()
      else:
          col_type = None
  ```
  **Problem:** The assignment of `None` to `col_type` if `col_type_class` is not found can be simplified.
  **Improvement:**
  ```python
  if isinstance(col_type, str):
      col_type_class = getattr(types, col_type, None)
      col_type = col_type_class() if col_type_class else None
  ```

### 3. General Review of Code Quality and Style

- **Code Quality:** The code changes are well-structured and address the problem of handling different column types effectively. The use of type hinting and type checking improves code readability and maintainability.
- **Code Style:** The code adheres to PEP 8 standards, making it clean and easy to read. The use of modern Python features like type hinting and `:=` operator (walrus operator) is commendable.
- **Testing:** The addition of unit tests for the `where_latest_partition` method in both Presto and Trino engines ensures that the changes are tested and validated.

### Additional Suggestions

- **Documentation:** Consider adding docstrings to the modified methods to explain the purpose and functionality of the changes.
- **Error Handling:** The use of broad exception handling (`raise Exception`) could be improved by raising more specific exceptions to provide better context in case of errors.

Overall, the pull request is well-implemented and addresses the problem effectively. The suggested improvements are minor and aim to further enhance the code quality.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)

**Comment by admsev** [comment]
@gooroodev review

**Review by michael-s-molina — APPROVED** [review]

**Pull request merged**